src/core/lib/layer-manager.js

```diff
@@ -318,19 +350,87 @@ export default class LayerManager {
     return redraw;
   }
 
-  //
-  // DEPRECATED METHODS
-  //
+  // Rebuilds viewports from descriptors towards a certain window size
+  _rebuildViewportsFromViews({viewDescriptors, width, height}) {
+    const newViewports = viewDescriptors.map(viewDescriptor =>
+      // If a `Viewport` instance was supplied, use it, otherwise build it
+      viewDescriptor.viewport instanceof Viewport ?
+        viewDescriptor.viewport :
+        this._makeViewportFromViewDescriptor({viewDescriptor, width, height})
+    );
 
-  setViewport(viewport) {
-    log.deprecated('setViewport', 'setViewports');
-    this.setViewports([viewport]);
-    return this;
+    this.setNeedsRedraw('Viewport(s) changed');
+
+    // Ensure one viewport is activated, layers may expect it
+    // TODO - handle empty viewport list (using dummy viewport), or assert
+    // const oldViewports = this.context.viewports;
+    // if (viewportsChanged) {
+
+    const viewport = newViewports[0];
+    assert(viewport instanceof Viewport, 'Invalid viewport');
+
+    this.context.viewports = newViewports;
+    this._activateViewport(viewport);
+    // }
+
+    // We've just rebuilt the viewports to match the descriptors, so clear the flag
+    this.viewports = newViewports;
+    this.viewDescriptorsChanged = false;
   }
 
-  //
-  // PRIVATE METHODS
-  //
+  // Build a `Viewport` from a view descriptor
+  _makeViewportFromViewDescriptor({viewDescriptor, width, height}) {
+    // Get the type of the viewport
+    // TODO - default to WebMercator?
+    const {type: ViewportType, viewState} = viewDescriptor;
+
+    // Resolve relative viewport dimensions
+    // TODO - we need to have width and height available
+    const viewportDimensions = this._getViewDimensions({viewDescriptor});
+
+    // Create the viewport, giving preference to view state in `viewState`
+    return new ViewportType(Object.assign({},
+      viewDescriptor,
+      viewportDimensions,
+      viewState // Object.assign handles undefined
+    ));
+  }
+
+  // Check if viewport array has changed, returns true if any change
+  // Note that descriptors can be the same
+  _diffViews(newViews, oldViews) {
+    if (newViews.length !== oldViews.length) {
+      return true;
+    }
+
+    return newViews.some(
+      (_, i) => this._diffView(newViews[i], oldViews[i])
+    );
+  }
+
+  // TODO: this will be true always if descriptors are used as we create new object
+  _diffView(newView, oldView) {
+    // `View` hiearchy supports an `equals` method
+    if (newView.viewport) {
+      return !oldView.viewport || !newView.viewport.equals(oldView.viewport);
+    }
+    return newView !== oldView;
+  }
+
+  // Support for relative viewport dimensions (e.g {y: '50%', height: '50%'})
+  _getViewDimensions({viewDescriptor, width, height}) {
+    const parsePercent = (value, max) => value;
```

Comment:
Is this intentionally left incomplete ?, Add a TODO for whats remaining and re-enable the code.